src/connectionController.ts

```diff
@@ -116,142 +128,206 @@ export default class ConnectionController {
-      log.error(
-        'Getting connection info with secrets failed because VSCode extension keytar module is undefined'
+      log.error('Could not migrate Keytar secrets, module not found');
+      return await this._storageController.saveConnection<MigratedStoreConnectionInfoWithConnectionOptions>(
+        {
+          ...savedConnectionInfo,
+          secretStorageLocation: SecretStorageLocation.Keytar,
+        }
       );
-      return savedConnectionInfo as StoreConnectionInfoWithConnectionOptions;
     }
 
-    try {
-      const unparsedSecrets = await ext.keytarModule.getPassword(
+    // If there is nothing in keytar, we will save an empty object as secrets in
+    // new storage and mark this connection as migrated
+    const keytarSecrets =
+      (await ext.keytarModule.getPassword(
         this._serviceName,
         savedConnectionInfo.id
-      );
-
-      // Ignore empty secrets.
-      if (!unparsedSecrets) {
-        return savedConnectionInfo as StoreConnectionInfoWithConnectionOptions;
-      }
+      )) || '{}';
 
-      const secrets = JSON.parse(unparsedSecrets);
-      const connectionOptions = savedConnectionInfo.connectionOptions;
-      const connectionInfoWithSecrets = mergeSecrets(
+    const migratedConnectionInfoWithSecrets =
+      this._mergedConnectionInfoWithSecrets(
         {
-          id: savedConnectionInfo.id,
-          connectionOptions,
-        } as ConnectionInfo,
-        secrets
+          ...savedConnectionInfo,
+          secretStorageLocation: SecretStorageLocation.SecretStorage,
+        },
+        keytarSecrets
       );
 
-      return {
-        ...savedConnectionInfo,
-        connectionOptions: connectionInfoWithSecrets.connectionOptions,
-      };
-    } catch (error) {
-      // Here we're lenient when loading connections in case their
-      // connections have become corrupted.
-      log.error('Getting connection info with secrets failed', error);
-      return;
-    }
-  }
-
-  private async _loadSavedConnectionsByStore(
-    savedConnections: ConnectionsFromStorage
-  ): Promise<void> {
-    if (!savedConnections || !Object.keys(savedConnections).length) {
-      return;
-    }
-
-    /** User connections are being saved both in:
-     * 1. Vscode global/workspace storage (without secrets) + keychain (secrets)
-     * 2. Memory of the extension (with secrets)
-     */
-    await Promise.all(
-      Object.keys(savedConnections).map(async (connectionId) => {
-        // Get connection info from vscode storage and merge with secrets.
-        const connectionInfoWithSecrets =
-          await this._getConnectionInfoWithSecrets(
-            savedConnections[connectionId]
-          );
-
-        // Save connection info with secrets to extension memory.
-        if (connectionInfoWithSecrets) {
-          this._connections[connectionId] = connectionInfoWithSecrets;
-        }
+    await this._saveConnectionWithSecrets(migratedConnectionInfoWithSecrets);
```

Comment:
When this succeeds, should we (or do we already?) also remove the keytar password?

Reply:
We talked about this and agreed to keep keytar password in case users try to use an older version of vscode, that is very unlikely, but still might be the case (and was useful for debugging this PR :)). We clean those secrets when a user deletes a connection, then secrets are deleted both from vscode secret storage and keytar.

Reply:
Earlier I was doing that but in this iteration I am not because I just thought that it would then not be an explicit user action which I tried avoiding. An argument against it could be to not leave orphan data behind. While writing this I started leaning a little more towards removing the secret from keytar as well. Let me know if we all think that is what we should do.

Reply:
I'm cool with either 👍 
Mostly wanted to check it was something we considered.